Add 2.9.0 release notes (#8994)
PrefectHQ/prefect: RELEASE-NOTES.md

        {'repo': 'rainbow-bridge', 'language': 'JavaScript', 'stars': 170000},
        {'repo': 'bubble-sort', 'language': 'JavaScript', 'stars': 184000}
    ]

    create_table_artifact(
        key="fun-data-table",
        table=my_table,
        description= "Our most heavily utilized repos for this week!"
    )

if __name__ == "__main__":
    my_fn()
```

You can view your artifacts in the Artifacts page of the Prefect UI, easily search your the data in your new table artifact, and toggle between a rendered and raw version of your data.

<insert image>

### Configure result storage keys

When persisting results, Prefect stores data at a unique, randomly-generated path. While this is convenient for ensuring the result is never overwritten, it limits organization of result files. In this release, we've added configuration of result storage keys, which gives you control over the result file path. Result storage keys can be dynamically formatted with access to all of the modules in `prefect.runtime` and the run's `parameters`. 

For example, you can name each result to correspond to the flow run that produced it and a parameter it received:

> ```

Reviewer: ```suggestion
```python
```

### Documentation
- Disambiguate reference to "Blocks" - https://github.com/PrefectHQ/prefect/pull/8921
- Fix broken concepts link - https://github.com/PrefectHQ/prefect/pull/8923
- Add note about fine-grained PAT format - https://github.com/PrefectHQ/prefect/pull/8929
- Add `UnpersistedResult` type - https://github.com/PrefectHQ/prefect/pull/8953
- Update docs CSS and config for versioning compatibility - https://github.com/PrefectHQ/prefect/pull/8957
- Clarify Filesystem package dependencies - https://github.com/PrefectHQ/prefect/pull/8989
- Update flow runs documentation - https://github.com/PrefectHQ/prefect/pull/8919
- Fix missing backticks on Work Pools concept page - https://github.com/PrefectHQ/prefect/pull/8942
- Update links to the release notes in the installation guide - https://github.com/PrefectHQ/prefect/pull/8974

### Contributors
- @andreadistefano made their first contribution in https://github.com/PrefectHQ/prefect/pull/8942
- @knl made their first contribution in https://github.com/PrefectHQ/prefect/pull/8974
- @thomas-te made their first contribution in https://github.com/PrefectHQ/prefect/pull/8959

> 
> 
> 

Reviewer: ```suggestion

```

# Prefect Release Notes

## Release 2.9.0

### Track and manage artifacts

Most workflows produce or update an artifact of some kind, whether its a table, a file, or a model. With Prefect Artifacts, you can track changes to these outputs and richly display them in the UI as tables, markdown, and links. Artifacts may be associated with a particular task run, flow run, or even exist outside a flow run context, enabling you to not only observe your flows, but the objects that they interact with as well. 

<insert image>

A variety of artifact types are available. To create an artifact that produces a table, for example, you can use the `create_table_artifact()` function.

```python
from prefect.artifacts import create_table_artifact

def my_fn():
    my_table = [
        {'repo': 'deep-thoughts', 'language': 'Python', 'stars': 5200}, 
        {'repo': 'fizz-buzz', 'language': 'Python', 'stars': 31900},
        {'repo': 'rainbow-bridge', 'language': 'JavaScript', 'stars': 170000},
        {'repo': 'bubble-sort', 'language': 'JavaScript', 'stars': 184000}
    ]

    create_table_artifact(

>         key="fun-data-table",

Reviewer: ```suggestion
        key="popular-repos",
```

# Prefect Release Notes

## Release 2.9.0

### Track and manage artifacts

Most workflows produce or update an artifact of some kind, whether its a table, a file, or a model. With Prefect Artifacts, you can track changes to these outputs and richly display them in the UI as tables, markdown, and links. Artifacts may be associated with a particular task run, flow run, or even exist outside a flow run context, enabling you to not only observe your flows, but the objects that they interact with as well. 

<insert image>

A variety of artifact types are available. To create an artifact that produces a table, for example, you can use the `create_table_artifact()` function.

```python

> from prefect import flow, task

Reviewer: ```suggestion
```

# Prefect Release Notes

## Release 2.9.0

### Track and manage artifacts

> Most workflows produce or update an artifact of some kind, whether its a table, a file, or a model. With Prefect Artifacts, you can track changes to these outputs and richly display them in the UI as tables, markdown, and links. Artifacts may be associated with a particular task run, flow run, or even exist outside a flow run context, enabling you to not only observe your flows, but the objects that they interact with as well. 

Reviewer: ```suggestion
Most workflows produce or update an artifact of some kind, whether it's a table, a file, or a model. With Prefect Artifacts, you can track changes to these outputs and richly display them in the UI as tables, markdown, and links. Artifacts may be associated with a particular task run, flow run, or even exist outside a flow run context, enabling you to not only observe your flows, but the objects that they interact with as well. 
```

![Artifacts top-level view](https://user-images.githubusercontent.com/27291717/228905742-0bad7874-6b6b-4000-9111-1c4d0e0bd6e1.png)

A variety of artifact types are available. To create an artifact that produces a table, for example, you can use the `create_table_artifact()` function.

```python
from prefect.artifacts import create_table_artifact

def my_fn():
    my_table = [
        {'repo': 'deep-thoughts', 'language': 'Python', 'stars': 5200}, 
        {'repo': 'fizz-buzz', 'language': 'Python', 'stars': 31900},
        {'repo': 'rainbow-bridge', 'language': 'JavaScript', 'stars': 170000},
        {'repo': 'bubble-sort', 'language': 'JavaScript', 'stars': 184000}
    ]

    create_table_artifact(
        key="popular-repos",
        table=my_table,
        description= "Our most heavily utilized repos for this week!"
    )

if __name__ == "__main__":
    my_fn()
```

> You can view your artifacts in the Artifacts page of the Prefect UI, easily search your the data in your new table artifact, and toggle between a rendered and raw version of your data.

Reviewer: ```suggestion
You can view your artifacts in the Artifacts page of the Prefect UI, easily search the data in your new table artifact, and toggle between a rendered and raw version of your data.
```

![Table artifact in a timeline view](https://user-images.githubusercontent.com/27291717/228905740-bd297de9-6381-45ec-aba3-8b72def70a08.png)

> See the following pull requests for implementation details:

Reviewer: ```suggestion
See [the documentation](https://docs.prefect.io/concepts/results/#artifacts) for more information.

See the following pull requests for implementation details:
```

A variety of artifact types are available. To create an artifact that produces a table, for example, you can use the `create_table_artifact()` function.

```python
from prefect.artifacts import create_table_artifact

def my_fn():
    my_table = [
        {'repo': 'deep-thoughts', 'language': 'Python', 'stars': 5200}, 
        {'repo': 'fizz-buzz', 'language': 'Python', 'stars': 31900},
        {'repo': 'rainbow-bridge', 'language': 'JavaScript', 'stars': 170000},
        {'repo': 'bubble-sort', 'language': 'JavaScript', 'stars': 184000}
    ]

    create_table_artifact(
        key="popular-repos",
        table=my_table,
        description= "Our most heavily utilized repos for this week!"
    )

if __name__ == "__main__":
    my_fn()
```

You can view your artifacts in the Artifacts page of the Prefect UI, easily search the data in your new table artifact, and toggle between a rendered and raw version of your data.

![Table artifact in a timeline view](https://user-images.githubusercontent.com/27291717/228905740-bd297de9-6381-45ec-aba3-8b72def70a08.png)

> See [the documentation](https://docs.prefect.io/concepts/artifacts) for more information.
> 
> See the following pull requests for implementation details:

Reviewer: ```suggestion
See [the documentation](https://docs.prefect.io/concepts/artifacts) for more information, as well as the following pull requests for implementation details:
```

# Prefect Release Notes

## Release 2.9.0

### Track and manage artifacts

Most workflows produce or update an artifact of some kind, whether its a table, a file, or a model. With Prefect Artifacts, you can track changes to these outputs and richly display them in the UI as tables, markdown, and links. Artifacts may be associated with a particular task run, flow run, or even exist outside a flow run context, enabling you to not only observe your flows, but the objects that they interact with as well. 

<insert image>

### Result Storage Keys

> While Prefect allowed you to persist results to a storage block, the path could only be a UUID, so you were limited in how those results could be organized. This release introduces result storage keys, which enable you to configure the path of the result file in the result storage. Result storage keys can be formatted with access to all of the modules in `prefect.runtime` and the run's `parameters`. For example, you can name each result to correspond to the flow that produced it:

Reviewer 2: I think I'd say something like..

```suggestion
When persisting results, Prefect stores data at a unique randomly generated path. While this is convenient for ensuring the result is never overwritten, it limits organization of result files. In this release, we've added configuration of result storage keys, which gives you control over the result file path. Result storage keys can be dynamically formatted with access to all of the modules in `prefect.runtime` and the run's `parameters`. 
```

Worth noting this doesn't actually template anything from the parent flow in the below example so the previous example intro didn't make sense. We can make it a new paragraph.

```

> from prefect import flow, task 
> 
> @flow() 
> def my_flow(): 
>      hello_world() 
>      hello_world(name="foo") 
>      hello_world(name="bar") 
> 
> @task(persist_result=True, result_storage_key="hello-{parameters[name]}.json") 
> def hello_world(name: str = "world"): 
>      return f"hello {name}"
> 
> my_flow()

Reviewer 2: ```suggestion
from prefect import flow, task

@flow()
def my_flow():
    hello_world()
    hello_world(name="foo")
    hello_world(name="bar")

@task(
    persist_result=True,
    result_storage_key="hello__{flow_run.name}__{parameters[name]}.json",
)
def hello_world(name: str = "world"):
    return f"hello {name}"

my_flow()
```

```

Which will persist three result files in the storage directory:
```

> $ ls ~/.prefect/storage | grep "hello-" 
> hello-bar.json 
> hello-foo.json 
> hello-world.json

Reviewer 2: ```suggestion
$ ls ~/.prefect/storage | grep "hello__" 
hello__rousing-mushroom__bar.json
hello__rousing-mushroom__foo.json
hello__rousing-mushroom__world.json
```

```

See [the docs](https://docs.prefect.io/concepts/results/?h=result#result-storage-key) for more information. 

### Enhancements
- Add unique integers to worker thread names for inspection - https://github.com/PrefectHQ/prefect/pull/8908
- Add support for custom static task storage keys - https://github.com/PrefectHQ/prefect/pull/8924
- Add flow name, flow run name, and flow id to `prefect.runtime.flow_run`, enabling them to be retrieved during a flow run's execution - https://github.com/PrefectHQ/prefect/pull/8947, https://github.com/PrefectHQ/prefect/pull/8948, and https://github.com/PrefectHQ/prefect/pull/8951
- Add support to `JSONSerializer` for serialization of exceptions so they are persisted even on failure - https://github.com/PrefectHQ/prefect/pull/8922
- Add Gzip middleware to the UI and API FastAPI apps for compressing responses - https://github.com/PrefectHQ/prefect/pull/8931

> 

Reviewer 2: Should also note that we added `prefect.runtime.task_run`

### Fixes
- Fix imports in copytree backport for Python 3.7 - https://github.com/PrefectHQ/prefect/pull/8925
- Retry on sqlite operational errors - https://github.com/PrefectHQ/prefect/pull/8950

> - Use dot notation for helm release tag - https://github.com/PrefectHQ/prefect/pull/8911

Reviewer 2: Not sure this one belongs here

- https://github.com/PrefectHQ/prefect/pull/8947
- https://github.com/PrefectHQ/prefect/pull/8948
- https://github.com/PrefectHQ/prefect/pull/8949
- https://github.com/PrefectHQ/prefect/pull/8951
- https://github.com/PrefectHQ/prefect/pull/8954
- https://github.com/PrefectHQ/prefect/pull/8956

### Enhancements
- Add unique integers to worker thread names for inspection - https://github.com/PrefectHQ/prefect/pull/8908

> - Update the runtime to detect flow run information from task run contexts — https://github.com/PrefectHQ/prefect/pull/8951

Reviewer 2: @serinamarie sorry, but this one is different and belongs here still as it's not covered in the highlight

- Update the runtime to detect flow run information from task run contexts — https://github.com/PrefectHQ/prefect/pull/8951
- Add support to `JSONSerializer` for serialization of exceptions so they are persisted even on failure - https://github.com/PrefectHQ/prefect/pull/8922
- Add Gzip middleware to the UI and API FastAPI apps for compressing responses - https://github.com/PrefectHQ/prefect/pull/8931
- Add task_name to the `task_run` runtime module — https://github.com/PrefectHQ/prefect/pull/8954

> - Add flow run parameter retrieval to `flow_run` runtime module — https://github.com/PrefectHQ/prefect/pull/8956

Reviewer 2: These can be linked in the above highlight on runtime changes instead
